Fix sync to genesis after upgrade from 1.4.x to 1.5

alsrdn opened the pull request:
* Allow for a sync leap to be returned for era 0 if the it contains the genesis block. Even though the sync leap does not contain a switch block, the block headers will still be included in the response.
* Infer validator weights for era 0 when registering the weights for era 1 since they should be the same given that the weights don't change until the unbonding delay elapsed.
* Change the nctl upgrade test to wait and check that the nodes have synced to genesis.

Fixes: https://github.com/casper-network/casper-node/issues/3631

alsrdn committed:
nctl/upgrade tests: wait for nodes to sync to genesis in scenario 1

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
validator matrix: add method to infer validator weights for era 0

Because of the unbonding delay, validator weights for era 0 are the same
as the ones from era 1. So when registering the weights for era 1, we
can also populate the validator matrix with the same weights for era 0.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
historical sync: fix leap to era 0 of pre 1.5 network

In a pre 1.5 network, we can't get the trusted block headers from era 0
because the sync leap is expected to return the switch block of the
previous era in the sync leap response but it won't if the trusted block
is part of era 0 of a pre 1.5 network where we don't have an immediate
switch block after genesis.

Allow a sync leap response to return the blocks in era 0 if genesis is
in this era.
Also use the method to infer era 0 validators when registering era 1.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

rafal-ch commented on node/src/types/validator_matrix.rs:
```diff
@@ -92,6 +92,27 @@ impl ValidatorMatrix {
+    ) -> bool {
+        let was_present = self.register_era_validator_weights(validators.clone());
+        if validators.era_id() == EraId::from(1) {
+            self.register_era_validator_weights(EraValidatorWeights::new(
```
Can you add an "info" or "debug" level log here, saying "infering validator weights for era 0"?

alsrdn replied:
Info level seems a bit excessive I think but a debug log is a good ideea.

alsrdn replied:
added in https://github.com/casper-network/casper-node/pull/3639/commits/7a7a7adc38ff918e87fc1265f7bef868149c1b54

rafal-ch commented on utils/nctl/sh/scenarios-upgrades/upgrade_scenario_01.sh:
```diff
@@ -301,6 +301,25 @@ function _step_08()
         fi
     done
+
+    log "Waiting for all nodes to sync to genesis"
+    TIMEOUT=300
+    for NODE_ID in $(seq 1 "$(get_count_of_nodes)")
+    do
+        log "Waiting for node $NODE_ID to sync to genesis (timeout=$TIMEOUT sec)"
+        RETRY_COUNT=1
+        LOWEST_AVAILABLE_BLOCK=$(get_node_lowest_available_block "$NODE_ID")
+        while [ -z "$LOWEST_AVAILABLE_BLOCK" ] || [ "$LOWEST_AVAILABLE_BLOCK" -ne 0 ]; do
+            if [ "$RETRY_COUNT" -gt $TIMEOUT ]; then
+                log "ERROR :: NODE-$NODE_ID RETRY :: Failed to sync to genesis within $TIMEOUT seconds; lowest available block $LOWEST_AVAILABLE_BLOCK"
+                exit 1
+            else
+                LOWEST_AVAILABLE_BLOCK=$(get_node_lowest_available_block "$NODE_ID")
+                sleep 1
+                ((RETRY_COUNT=RETRY_COUNT+1))
+            fi
+        done
+    done
```
What do you think about extracting [do_await_node_1_historical_sync](https://github.com/casper-network/casper-node/blob/749cab93492a9f59c161c152732560e213511540/utils/nctl/sh/scenarios/upgrade_after_emergency_upgrade_test.sh#L200), change it to `do_await_node_historical_sync(node_num)` and use here?

rafal-ch replied:
Actually, if you choose to do the above, you can then easily add this check to all of the upgrade tests, potentially saving us some trouble :)

alsrdn replied:
That's a good point, didn't see that we had that similar function. Will change.

alsrdn replied:
added in https://github.com/casper-network/casper-node/pull/3639/commits/e02e09277d030a9e81736eadb34d7e81aeb6aac9

alsrdn committed:
validator matrix: add debug message when infering era 0 weights

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
nctl: add common function for waiting for sync to genesis

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn commented:
bors try

georgepisaltu commented on node/src/reactor/main_reactor/keep_up.rs:
```diff
@@ -667,18 +670,24 @@ impl MainReactor {
                             // for the earlier era's proof
-                            Some(previous_era_id) => Ok(Some(SyncBackInstruction::Sync {
+                            block_header.era_id().predecessor()
+                        };
```
This code block now never returns `None`, so we can ditch the `Option` indirection. We can make this even simpler by doing `block_header.era_id().saturating_sub(1)`.

alsrdn replied:
Changed in https://github.com/casper-network/casper-node/pull/3639/commits/260398e7cfb0a635682911a10d2106b3d5375ec1

georgepisaltu reviewed (commented):
LGTM overall

rafal-ch reviewed (approved):
LGTM :+1:

alsrdn committed:
keep_up: simplify sync back leap era id selection

Simplify code for era id selection when the trusted block header is not
in storage.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn commented:
bors merge
